plugins/modem-manager/fu-qmi-pdc-updater.c

```diff
@@ -0,0 +1,432 @@
+/*
+ * Copyright (C) 2019 Aleksander Morgado <aleksander@aleksander.es>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include "fu-qmi-pdc-updater.h"
+
+#define FU_QMI_PDC_MAX_OPEN_ATTEMPTS 8
+
+struct _FuQmiPdcUpdater {
+	GObject		parent_instance;
+	gchar		*qmi_port;
+	QmiDevice	*qmi_device;
+	QmiClientPdc	*qmi_client;
+};
+
+G_DEFINE_TYPE (FuQmiPdcUpdater, fu_qmi_pdc_updater, G_TYPE_OBJECT)
+
+typedef struct {
+	GMainLoop	*mainloop;
+	QmiDevice	*qmi_device;
+	QmiClientPdc	*qmi_client;
+	GError		*error;
+	guint		open_attempts;
+} OpenContext;
+
+static void fu_qmi_pdc_updater_qmi_device_open_attempt (OpenContext *ctx);
+
+static void
+fu_qmi_pdc_updater_qmi_device_open_abort_ready (GObject *qmi_device, GAsyncResult *res, gpointer user_data)
+{
+	OpenContext *ctx = (OpenContext *) user_data;
+
+	g_assert (ctx->error != NULL);
+
+	/* ignore errors when aborting open */
+	qmi_device_close_finish (QMI_DEVICE (qmi_device), res, NULL);
+
+	ctx->open_attempts--;
+	if (ctx->open_attempts == 0) {
+		g_clear_object (&ctx->qmi_client);
+		g_clear_object (&ctx->qmi_device);
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	/* retry */
+	g_clear_error (&ctx->error);
+	fu_qmi_pdc_updater_qmi_device_open_attempt (ctx);
+}
+
+static void
+fu_qmi_pdc_updater_open_abort (OpenContext *ctx)
+{
+	qmi_device_close_async (ctx->qmi_device,
+				15, NULL, fu_qmi_pdc_updater_qmi_device_open_abort_ready, ctx);
+}
+
+static void
+fu_qmi_pdc_updater_qmi_device_allocate_client_ready (GObject *qmi_device, GAsyncResult *res, gpointer user_data)
+{
+	OpenContext *ctx = (OpenContext *) user_data;
+
+	ctx->qmi_client = QMI_CLIENT_PDC (qmi_device_allocate_client_finish (QMI_DEVICE (qmi_device), res, &ctx->error));
+	if (ctx->qmi_client == NULL) {
+		fu_qmi_pdc_updater_open_abort (ctx);
+		return;
+	}
+
+	g_main_loop_quit (ctx->mainloop);
+}
+
+static void
+fu_qmi_pdc_updater_qmi_device_open_ready (GObject *qmi_device, GAsyncResult *res, gpointer user_data)
+{
+	OpenContext *ctx = (OpenContext *) user_data;
+
+	if (!qmi_device_open_finish (QMI_DEVICE (qmi_device), res, &ctx->error)) {
+		fu_qmi_pdc_updater_open_abort (ctx);
+		return;
+	}
+
+	qmi_device_allocate_client (ctx->qmi_device, QMI_SERVICE_PDC, QMI_CID_NONE, 5, NULL,
+				    fu_qmi_pdc_updater_qmi_device_allocate_client_ready, ctx);
+}
+
+static void
+fu_qmi_pdc_updater_qmi_device_open_attempt (OpenContext *ctx)
+{
+	QmiDeviceOpenFlags open_flags = QMI_DEVICE_OPEN_FLAGS_NONE;
+
+	g_debug ("trying to open QMI device...");
```

Comment:
maybe move down closer to qmi_device_open?

Reply:
ok